src/browser/transports/injection_bus.ts

```diff
@@ -88,28 +91,84 @@ export default class NativeWindowInjectionBus extends EventEmitter {
         return;
       }
 
-      // Ack / Nack
+      this._listenerHeartbeat(parsedMessage);
+
       if (this._pendingRequests.has(messageId)) {
-        const isNack = parsedMessage.action.includes('error');
-        const pendingRequest = this._pendingRequests.get(messageId);
+        this._listenerAck(parsedMessage);
+      } else {
+        this._listenerBroadcastMsg(parsedMessage);
+      }
+    };
 
-        isNack
-          ? pendingRequest.reject((<NackMessage>parsedMessage).payload.reason)
-          : pendingRequest.resolve();
+    copyDataTransport.on('message', this._messageListener);
+
+    this.subscribe();
+  }
 
-        this._senderId = senderId;
-        this._pendingRequests.delete(messageId);
+  // Setup heartbeat
+  private setupHeartbeat() {
+    this._connected = false;
+    this._previousDllSequence = -1;
 
-        return;
+    this._listenerHeartbeat = (message: MessageBase): void => {
+      const { sequence } = message;
+
+      if (this._previousDllSequence + 1 !== sequence) {
+        // At least one message has been missed
+        const msg = `[NWI] Missed message(s) from ${this._senderId}: `
+          + `previous sequence ${this._previousDllSequence}, `
+          + `current sequence: ${sequence}`;
+        writeToLog('info', msg);
+      }
+
+      this._previousDllSequence = sequence;
+    };
+
+    // Ping message window every second and update current connection status
+    setInterval(() => {
```

Comment:
This is not tied to the lifetime of a window? If so the interval needs to be cleaned up or its a leak

Reply:
Great catch 👍